Add 4.5-beta1 blog post (#1082)
godotengine/godot-website: collections/_article/dev-snapshot-godot-4-5-beta-1.md

---
title: "Dev snapshot: Godot 4.5 beta 1"
excerpt: Godot 4.5 has entered beta!
categories: [pre-release]
author: Thaddeus Crews
image: /storage/blog/covers/dev-snapshot-godot-4-5-beta-1.webp
image_caption_title: Rift Riff
image_caption_description: A game by Adriaan de Jongh
date: 2025-06-16 12:00:00
---

> The first beta release for the 4.5 release cycle has come at last, and with it a plethora of outstanding bugs to be squashed. Contributors are encouraged to focus exclusively on fixing [regressions](https://github.com/godotengine/godot/issues?q=is%3Aopen+is%3Aissue+label%3Aregression+milestone%3A4.5), as we are now in feature-freeze & will not be merging any new features at this stage of development.

Reviewer: ```suggestion
The first beta release for the 4.5 release cycle has come at last, and with it a plethora of outstanding bugs to be squashed. Contributors are encouraged to focus exclusively on fixing [regressions](https://github.com/godotengine/godot/issues?q=is%3Aopen+is%3Aissue+label%3Aregression+milestone%3A4.5), as we are now in feature-freeze and will not be merging any new features at this stage of development.
```

> Please, consider [supporting the project financially](#support), if you are able. Godot is maintained by the efforts of volunteers and a small team of paid contributors. Your donations go towards sponsoring their work and ensuring they can dedicate their undivided attention to the needs of the project.

Reviewer: ```suggestion
Please consider [supporting the project financially](#support), if you are able. Godot is maintained by the efforts of volunteers and a small team of paid contributors. Your donations go towards sponsoring their work and ensuring they can dedicate their undivided attention to the needs of the project.
```
I realise this is a piece of stock text, so might need editing elsewhere...

[Jump to the **Downloads** section](#downloads), and give it a spin right now, or continue reading to learn more about improvements in this release. You can also [try the **Web editor**](https://editor.godotengine.org/releases/4.5.beta1/) or the **Android editor** for this release. If you are interested in the latter, please request to join [our testing group](https://groups.google.com/g/godot-testers) to get access to pre-release builds.

---

> *The cover illustration is from* [**Rift Riff**](https://store.steampowered.com/app/2800900/Rift_Riff/?curator_clanid=41324400), *a tower defense game whose layered combat isn't compromised by its a variety of forgiving mechanics! You can buy the game [on Steam](https://store.steampowered.com/app/2800900/Rift_Riff/?curator_clanid=41324400), and follow the developer on [Bluesky](https://bsky.app/profile/adriaan.games).*

Reviewer: ```suggestion
*The cover illustration is from* [**Rift Riff**](https://store.steampowered.com/app/2800900/Rift_Riff/?curator_clanid=41324400), *a tower defense game whose layered combat isn't compromised by its variety of forgiving mechanics! You can buy the game [on Steam](https://store.steampowered.com/app/2800900/Rift_Riff/?curator_clanid=41324400), and follow the developer on [Bluesky](https://bsky.app/profile/adriaan.games).*
```
Although I'm not quite sure what that sentence means!

Reviewer 2: It's implying that the game is surprisingly lenient, despite how in-depth and complicated it can be ("whose layered combat").

## Highlights

- [New in Beta 1!](#new-in-beta-1)
- [Breaking changes](#breaking-changes)
- [Core](#core)
- [Editor](#editor)
- [GDScript](#gdscript)
- [GUI](#gui)
- [Internationalization](#internationalization)
- [Navigation](#navigation)
- [Physics](#physics)
- [Platforms](#platforms)
- [Rendering and shaders](#rendering-and-shaders)

### New in Beta 1!

> Much like our [4.4 beta1 blog post](/article/dev-snapshot-godot-4-4-beta-1/#new-in-beta-1), we have many new features to showcase that managed to squeeze in before the feature freeze! Similarly, most of these are somewhat experimental compared to the more traditional highlights later on in the article, so please let us know if any issues are encountered.

Reviewer: ```suggestion
Much like our [4.4 beta 1 blog post](/article/dev-snapshot-godot-4-4-beta-1/#new-in-beta-1), we have many new features to showcase that managed to squeeze in before the feature freeze! Similarly, most of these are somewhat experimental compared to the more traditional highlights later on in the article, so please let us know if any issues are encountered.
```

#### Stencil support for spatial materials
With a PR in the making for nearly 2 years [GH-80710](https://github.com/godotengine/godot/pull/80710), [apples](https://github.com/apples) delivers the long-awaited stencil support to Godot!

|                                                      Before                                                       |                                                      After                                                      |
| :---------------------------------------------------------------------------------------------------------------: | :-------------------------------------------------------------------------------------------------------------: |
| <img src="/storage/blog/dev-snapshot-godot-4-5-beta-1/stencil-before-1.webp" alt="Stencil before 1" width="350"/> | <img src="/storage/blog/dev-snapshot-godot-4-5-beta-1/stencil-after-1.webp" alt="Stencil after 1" width="350"/> |
| <img src="/storage/blog/dev-snapshot-godot-4-5-beta-1/stencil-before-2.webp" alt="Stencil before 2" width="350"/> | <img src="/storage/blog/dev-snapshot-godot-4-5-beta-1/stencil-after-2.webp" alt="Stencil after 2" width="350"/> |

> 

Reviewer: Looking at the PR comments, I don't think those are before and after pics.

Author: Ahh, you're right. I saw a table and autopiloted. Gonna just use the screengrabs from the OP instead

#### GDScript: Abstract and variadic functions

> GDScript actually saw a significant number of high-profile merges for beta1: [autocompletion for user methods](https://github.com/godotengine/godot/pull/106198), [highlighting script members like native ones](https://github.com/godotengine/godot/pull/74393), and [autocompletion for `@export_tool_button`](https://github.com/godotengine/godot/pull/105081) just to name a few! Even for this highlight we couldn't narrow it down to just one, so focus will be put on two that achieve similar goals of long-awaited additions to method functionality: abstract functions & variadic functions.

Reviewer: ```suggestion
GDScript saw a significant number of high-profile merges for beta 1: [autocompletion for user methods](https://github.com/godotengine/godot/pull/106198), [highlighting script members like native ones](https://github.com/godotengine/godot/pull/74393), and [autocompletion for `@export_tool_button`](https://github.com/godotengine/godot/pull/105081) just to name a few! Even for this highlight we couldn't narrow it down to just one, so focus will be put on two that achieve similar goals of long-awaited additions to method functionality: abstract functions and variadic functions.
```

> The abstract method PR [GH-106409](https://github.com/godotengine/godot/pull/106409) comes courtesy of [Danil Alexeev](https://github.com/godotengine/godot/pull/106409), and acts as a direct continuation to the abtract classes introduced in our [previous blog post](/article/dev-snapshot-godot-4-5-dev-5/#gdscript-abstract-classes). By prepending the `abstract` keyword to a function, it will be marked for explicit override by child classes.

Reviewer: ```suggestion
The abstract method PR [GH-106409](https://github.com/godotengine/godot/pull/106409) comes courtesy of [Danil Alexeev](https://github.com/dalexeev), and acts as a direct continuation to the abstract classes introduced in our [previous blog post](/article/dev-snapshot-godot-4-5-dev-5/#gdscript-abstract-classes). By prepending the `abstract` keyword to a function, it will be marked for explicit override by child classes.
```

```gdscript
abstract class Item:
	abstract func get_name() -> String

	func use() -> void:
		print("Character used %s." % get_name())

class HealingPotion extends Item:
	func get_name() -> String:
		return "Healing Potion"

func _ready() -> void:
	var potion := HealingPotion.new()
	potion.use() # Prints `Character used Healing Potion.`
```

> Danil isn't done yet though, as the variadic PR [GH-82808](https://github.com/godotengine/godot/pull/82808) comes from him as well! A variadic argument means that the argument itself is optional, the final argument in a given function, and accepts a sequence of arguments as if there were multiple argument slots beyond it.

Reviewer: ?

```gdscript
func f(a: int, b: int = 0, ...args: Array):
	prints(a, b, args)

func _ready() -> void:
	f(1)
	f(1, 2)
	f(1, 2, 3)
	f(1, 2, 3, 4)
	f(1, 2, 3, 4, 5)

# Output:
# 1 0 []
# 1 2 []
# 1 2 [3]
# 1 2 [3, 4]
# 1 2 [3, 4, 5]
```

<img src="/storage/blog/dev-snapshot-godot-4-5-beta-1/variadic-documentation.webp" alt="Variadic documentation"/>

### Breaking changes
We try to minimize breaking changes, but sometimes they are necessary in order to fix high priority issues. Where we do break compatibility, we do our best to make sure that the changes are minimal and require few changes in user projects.

You can find a list of such issues by filtering the merged PRs in the 4.5 milestone with the [`breaks compat` label](https://github.com/godotengine/godot/issues?q=milestone%3A4.5%20is%3Amerged%20label%3A%22breaks%20compat%22). Here are some which are worth being aware of:

- TODO

### Core

> Changes to the core of the engine require significantly more scrutiny than other parts of the engine; this comes down to how critical and foundational virtually every single piece of code proves to be. This makes it all the more impressive that [bruvzg](https://github.com/bruvzg) managed to implement [AccessKit](https://github.com/AccessKit/accesskit) support in such an integral manner ([GH-76829](https://github.com/godotengine/godot/pull/76829)). Additionally, special thanks to [Lukas Tenbrink](https://github.com/Ivorforce), a new addition to the core team who has been (contributing)[https://github.com/godotengine/godot/pull/103708] (nonstop)[https://github.com/godotengine/godot/pull/106996] (improvements)[https://github.com/godotengine/godot/pull/104381] to ensure optimal performance for developers and maintainers alike.

Reviewer: ```suggestion
Changes to the core of the engine require significantly more scrutiny than other parts of the engine; this comes down to how critical and foundational virtually every single piece of code proves to be. This makes it all the more impressive that [bruvzg](https://github.com/bruvzg) managed to implement [AccessKit](https://github.com/AccessKit/accesskit) support in such an integral manner ([GH-76829](https://github.com/godotengine/godot/pull/76829)). Additionally, special thanks to [Lukas Tenbrink](https://github.com/Ivorforce), a new addition to the core team who has been [contributing](https://github.com/godotengine/godot/pull/103708) [non-stop](https://github.com/godotengine/godot/pull/106996) [improvements](https://github.com/godotengine/godot/pull/104381) to ensure optimal performance for developers and maintainers alike.
```

### Editor

> A rarely-covered topic regarding the editor is the [command palette](https://docs.godotengine.org/en/stable/classes/class_editorcommandpalette.html), but we'll happily make an exception to highlight [HolonProduction](https://github.com/HolonProduction)'s PR [GH-99318](https://github.com/godotengine/godot/pull/99318) adding named `EditorScript`s to the command palette! This much more centralized means of execution serves to benefit the commands that are more project-specific. This is specifically for named scripts however, such that there will always be an associated display name & search handling.

Reviewer: ```suggestion
A rarely-covered topic regarding the editor is the [command palette](https://docs.godotengine.org/en/stable/classes/class_editorcommandpalette.html), but we'll happily make an exception to highlight [HolonProduction](https://github.com/HolonProduction)'s PR [GH-99318](https://github.com/godotengine/godot/pull/99318) adding named `EditorScript`s to the command palette! This much more centralized means of execution serves to benefit the commands that are more project-specific. This is specifically for named scripts however, such that there will always be an associated display name and search handling.
```

### GDScript

> Adding backtracing to GDScript was among the most highly requested features from our users for years. By their powers combined, [Mikael Hermansson](https://github.com/mihe) and [Juan Linietsky](https://github.com/reduz) have added script backtracing support to GDScript ([GH-91006](https://github.com/godotengine/godot/pull/91006)). Finding the root problem behind warnings/errors required manually hunting down the issue with very little assistance from the editor itself; a problem that will no longer plague users that utilize the backtracing functionality. This functionality is always available in debug mode, but can be activated in release mode if **Debug > Settings > GDScript > Always Track Call Stacks** is enabled in the project settings. This can make it easier for users to report issues in a way that developers can track down.

Reviewer: ```suggestion
Adding backtracing to GDScript was among the most highly-requested features from our users for years. By their powers combined, [Mikael Hermansson](https://github.com/mihe) and [Juan Linietsky](https://github.com/reduz) have added script backtracing support to GDScript ([GH-91006](https://github.com/godotengine/godot/pull/91006)). Finding the root problem behind warnings/errors required manually hunting down the issue with very little assistance from the editor itself; a problem that will no longer plague users that utilize the backtracing functionality. This functionality is always available in debug mode, but can be activated in release mode if **Debug > Settings > GDScript > Always Track Call Stacks** is enabled in the project settings. This can make it easier for users to report issues in a way that developers can track down.
```

> As mentioned in the new features section, abstract methods were a continuation of abstract classes. While the ability to create abstract classes was always available internally, it's thanks [Aaron Franke](https://github.com/aaronfranke) that GDScript users can take advantage of this this paradigm is now available to GDScript users ([GH-67777](https://github.com/godotengine/godot/pull/67777)). Now if a user wants to introduce their own abstract class, they merely need to declare it via the new `abstract` keyword.

Reviewer: ```suggestion
As mentioned in the new features section, abstract methods were a continuation of abstract classes. While the ability to create abstract classes was always available internally, it's thanks to [Aaron Franke](https://github.com/aaronfranke) that GDScript users can now take advantage of this paradigm ([GH-67777](https://github.com/godotengine/godot/pull/67777)). Now if a user wants to introduce their own abstract class, they merely need to declare it via the new `abstract` keyword.
```

### GUI
4.5 is bringing with it quite a few new quality-of-life improvements, with arguably the biggest addition being one we actually haven't given proper coverage to yet: foldable containers! [Tomasz Chabora](https://github.com/KoBeWi) is no stranger to editor enhancements, and this time he has blessed us with [GH-102346](https://github.com/godotengine/godot/pull/102346), which grants us the new `FoldableContainer` class. Now users can have dynamically cascading GUI objects with the ability to toggle if the contents are expanded or not at will, a process that previously took several workarounds to achieve.

> In a similar vein, the ability to manipulate a group of controls simultaneously has become much easier thanks to [Delsin-Yu](https://github.com/Delsin-Yu)'s PR with a focus on recursive control across child controls ([GH-97495](https://github.com/godotengine/godot/pull/97495)). This implements new proprties for recursively disabling `Focus Mode` and `Mouse Filter`, meaning that the ability to select & interact with child controls becomes far more intuitive, and allows for explicit overrides if desired.

Reviewer: ```suggestion
In a similar vein, the ability to manipulate a group of controls simultaneously has become much easier thanks to [Delsin-Yu](https://github.com/Delsin-Yu)'s PR with a focus on recursive control across child controls ([GH-97495](https://github.com/godotengine/godot/pull/97495)). This implements new properties for recursively disabling `Focus Mode` and `Mouse Filter`, meaning that the ability to select and interact with child controls becomes far more intuitive, and allows for explicit overrides if desired.
```

<video autoplay loop muted playsinline>
  <source src="/storage/blog/godot-4-5-beta-1/recursive-focus.webm?1" type="video/webm">
</video>

> There've also been a few improvements to `SVGTexture` in 4.5, such as the [dedicated importer](https://github.com/godotengine/godot/pull/105655). However, as it pertains to GUI, the biggest improvement would be [bruvzg](https://github.com/bruvzg)'s implementation of auto-scalable SVGs ([GH-105375](https://github.com/godotengine/godot/pull/105375)). This will allow users to explicitly oversample fonts and raster graphics, resulting a much crisper image overall.

Reviewer: ```suggestion
There have also been a few improvements to `SVGTexture` in 4.5, such as the [dedicated importer](https://github.com/godotengine/godot/pull/105655). However, as it pertains to GUI, the biggest improvement would be [bruvzg](https://github.com/bruvzg)'s implementation of auto-scalable SVGs ([GH-105375](https://github.com/godotengine/godot/pull/105375)). This will allow users to explicitly oversample fonts and raster graphics, resulting a much crisper image overall.
```

<video autoplay loop muted playsinline>
  <source src="/storage/blog/godot-4-5-beta-1/svg-auto-scalable.webm?1" type="video/webm">
</video>

### Internationalization
Internationalization has always been an extremely crucial part of the Godot Project, and its relative lack of coverage in our blogposts does not do it justice. [Haoyu Qiu](https://github.com/timothyqiu) allows us to break this bad habit of ours with a fantastic addition to the i18n workflow: editor previews! ([GH-96921](https://github.com/godotengine/godot/pull/96921)) Now there's much less guesswork necessary for how your projects will appear to the end-user in their native language, ensuring a consistent and clean style no matter the selection!

> Similarly, the ability to swap languages on-the-fly within the editor is now possible thanks to the efforts of [Tomasz Chabora](https://github.com/KoBeWi) and [GH-102562](https://github.com/godotengine/godot/pull/102562). Now users can preview *and* experience multiple language options in a single editor session.

Reviewer: ```suggestion
Similarly, the ability to swap languages on-the-fly within the editor is now possible thanks to the efforts of [Tomasz Chabora](https://github.com/KoBeWi) in [GH-102562](https://github.com/godotengine/godot/pull/102562). Now users can preview *and* experience multiple language options in a single editor session.
```

### Navigation

> The navigation team has been able to really spread their wings thanks to logic for 2D and 3D being handled independantly. [smix8](https://github.com/smix8) took the mantle on the initial split, while [AThousandShips](https://github.com/AThousandShips) brought the logic to the module system itself. This opened the door for improvements to performance across both navigation systems, and *dramatically* decreased the size of builds which exclusively target 2D.

Reviewer: ```suggestion
The navigation team has been able to really spread their wings thanks to logic for 2D and 3D being handled independently. [smix8](https://github.com/smix8) took the mantle on the initial split, while [AThousandShips](https://github.com/AThousandShips) brought the logic to the module system itself. This opened the door for improvements to performance across both navigation systems, and *dramatically* decreased the size of builds which exclusively target 2D.
```

### Physics

> Our implementation of 3D interpolation has been completely overhauled, as the previous iteration had fundamental flaws with how it was structured in such a way that couldn't be addressed with a simple fix or patch. The solution already existed on 3.x, and has since been forward ported by [lawnjelly](https://github.com/lawnjelly) via [GH-104269](https://github.com/godotengine/godot/pull/104269). Now all logic is handled within the `SceneTree`, but in a manner which doesn't break any existing API. All projects will benefit from this improvement out-of-the-box.

Reviewer: ```suggestion
Our implementation of 3D interpolation has been completely overhauled, as the previous iteration had fundamental flaws with how it was structured in such a way that couldn't be addressed with a simple fix or patch. The solution already existed on 3.x, and has since been forward-ported by [lawnjelly](https://github.com/lawnjelly) via [GH-104269](https://github.com/godotengine/godot/pull/104269). Now all logic is handled within the `SceneTree`, but in a manner which doesn't break any existing API. All projects will benefit from this improvement out-of-the-box.
```

### Platforms
#### Android

> The Android editor experience has become significantly improved thanks to the implementation of `TouchActionsPanel`. While this is technically something any touch device can take advantage of, [Anish Mishra](https://github.com/syntaxerror247)'s PR [GH-100339](https://github.com/godotengine/godot/pull/100339) was explicitly created with the Android editor in mind. `TouchActionsPanel` comes equiped with common actions buttons (save, undo, redo, etc.), which simulate actions like `ui_undo` and `ui_redo` via pre-existing shortcuts. This was implemented early on in the 4.5 development cycle, giving our maintainers a stable and versitile foundation for future touch-based actions.

Reviewer: ```suggestion
The Android editor experience has become significantly improved thanks to the implementation of `TouchActionsPanel`. While this is technically something any touch device can take advantage of, [Anish Mishra](https://github.com/syntaxerror247)'s PR [GH-100339](https://github.com/godotengine/godot/pull/100339) was explicitly created with the Android editor in mind. `TouchActionsPanel` comes equipped with common action buttons (save, undo, redo, etc.), which simulate actions like `ui_undo` and `ui_redo` via pre-existing shortcuts. This was implemented early on in the 4.5 development cycle, giving our maintainers a stable and versatile foundation for future touch-based actions.
```

<video autoplay loop muted playsinline>
  <source src="/storage/blog/godot-4-5-beta-1/touch-actions-panel?1" type="video/webm">
</video>

#### macOS

> It took a while for embedded window support to come to macOS, as this OS does not allow the kind of window manipulation that Windows and Linux use for game window embedding. Instead, macOS utilizes an inter-process communication approach where the framebuffer is sent from the game process (which performs off-screen rendering) to the editor window, which also handles input events to the game process. Although it's much more complex, [Stuart Carnie](https://github.com/stuartcarnie) used a more robust approach doesn't rely on any window management hacks ([GH-105884](https://github.com/godotengine/godot/pull/105884)). Not having to worry about edge-cases or low-level shenanigans is appealing in its own right, so this approach may be ported later to Windows/Linux in a future release to make game window embedding much more reliable overall.

Reviewer: ```suggestion
It took a while for embedded window support to come to macOS, as this OS does not allow the kind of window manipulation that Windows and Linux use for game window embedding. Instead, macOS utilizes an inter-process communication approach where the framebuffer is sent from the game process (which performs off-screen rendering) to the editor window, which also handles input events to the game process. Although it's much more complex, [Stuart Carnie](https://github.com/stuartcarnie) used a more robust approach that doesn't rely on any window management hacks ([GH-105884](https://github.com/godotengine/godot/pull/105884)). Not having to worry about edge-cases or low-level shenanigans is appealing in its own right, so this approach may be ported later to Windows/Linux in a future release to make game window embedding much more reliable overall.
```

#### visionOS
The Apple XR environment, visionOS, comes to Godot Engine in 4.5! We're already committed to making great strides in XR featuresets and support, but this brings us closer to a platform-independent pipeline similar to what we already have with traditional platforms. It might be unconventional to imagine using an engine editor within an XR environment, but it's one which we've supported for Meta Quest and OpenXR; now Apple users can join the fun.

#### Web

> The upcoming performance boost from SIMD is so impressive that we've made a (dedicated article)[/article/upcoming-serious-web-performance-boost/] to showcase exactly what kind of benefits users can expect. The short version is [Adam Scott](https://github.com/adamscott) created a PR which changed only a single compiler flag ([GH-106319](https://github.com/godotengine/godot/pull/106319)), but one which caused universal improvement for web applications and editor workflow. Check out our article for the long version.

Reviewer: ```suggestion
The upcoming performance boost from SIMD is so impressive that we've made a [dedicated article](/article/upcoming-serious-web-performance-boost/) to showcase exactly what kind of benefits users can expect. The short version is that [Adam Scott](https://github.com/adamscott) created a PR which changed only a single compiler flag ([GH-106319](https://github.com/godotengine/godot/pull/106319)), but one which caused universal improvements for web applications and editor workflow. Check out our article for the long version.
```

| :------------------------------------------------------------------------------------------------------------------------: | :----------------------------------------------------------------------------------------------------------------------: |
| <img src="/storage/blog/dev-snapshot-godot-4-5-dev-5/bent-normals-before-1.webp" alt="Bent normals before 1" width="350"/> | <img src="/storage/blog/dev-snapshot-godot-4-5-dev-5/bent-normals-after-1.webp" alt="Bent normals after 1" width="350"/> |
| <img src="/storage/blog/dev-snapshot-godot-4-5-dev-5/bent-normals-before-2.webp" alt="Bent normals before 2" width="350"/> | <img src="/storage/blog/dev-snapshot-godot-4-5-dev-5/bent-normals-after-2.webp" alt="Bent normals after 2" width="350"/> |
| <img src="/storage/blog/dev-snapshot-godot-4-5-dev-5/bent-normals-before-3.webp" alt="Bent normals before 3" width="350"/> | <img src="/storage/blog/dev-snapshot-godot-4-5-dev-5/bent-normals-after-3.webp" alt="Bent normals after 3" width="350"/> |

> 

Reviewer: I wonder if there's a way to show these images so that users can click to switch between them? Maybe also allowing them to do so while zooming in. At the moment, it is hard to tell the difference from side-by-side images.

Reviewer 3: Click-to-change or a slider would be nice, but [here](https://github.com/godotengine/godot-docs/pull/10996) are zoomed-in images, where differences can be seen more easily.

---
title: "Dev snapshot: Godot 4.5 beta 1"
excerpt: Godot 4.5 has entered beta!
categories: [pre-release]
author: Thaddeus Crews
image: /storage/blog/covers/dev-snapshot-godot-4-5-beta-1.webp
image_caption_title: Rift Riff
image_caption_description: A game by Adriaan de Jongh
date: 2025-06-16 12:00:00
---

The first beta release for the 4.5 release cycle has come at last, and with it a plethora of outstanding bugs to be squashed. Contributors are encouraged to focus exclusively on fixing [regressions](https://github.com/godotengine/godot/issues?q=is%3Aopen+is%3Aissue+label%3Aregression+milestone%3A4.5), as we are now in feature-freeze and will not be merging any new features at this stage of development.

> For those interested in aiding us on our quest to squash any bugs that come up during this time, we encourage you to join our recent bug-hunting sprints. Helmed by our new head of Bugsquad, [AThousandShips](https://github.com/AThousandShips) has taken to hosting regular sprints for tackling various bugs within the Godot repo, organized such that everyone can easily gather behind a given theme to make chunks of fixes as seamless and speedy as possible. This will be our fourth sprint of this type, with an associated writeup [already prepared](https://gist.github.com/AThousandShips/73002da28f7c66ac8166c64c926e9cc7). See the [Bug Triage Introduction](https://github.com/godotengine/godot-maintainers-docs/blob/main/bug-triage/introduction.md) for more information!

Reviewer 4: Should link to the discussion directly IMO, as the gist is just for details and doesn't help participating if you don't have the discussion:
* https://chat.godotengine.org/channel/dNtaAGsF2ifAjZ57P

Should also list the bugsquad and bugsqud-sprints channels:
* https://chat.godotengine.org/channel/bugsquad
* https://chat.godotengine.org/channel/bugsquad-sprints

Reviewer 4: ```suggestion
For those interested in aiding us on our quest to squash any bugs that come up during this time, we encourage you to join our recent bug-hunting sprints. Helmed by our new head of the Bugsquad, [AThousandShips](https://github.com/AThousandShips) has taken to hosting regular sprints for tackling various bugs within the Godot repo, organized such that everyone can easily gather behind a given theme to make chunks of fixes as seamlessly and speedily as possible. This will be our fourth sprint of this type, with an associated writeup [already prepared](https://gist.github.com/AThousandShips/73002da28f7c66ac8166c64c926e9cc7). See the [Bug Triage Introduction](https://github.com/godotengine/godot-maintainers-docs/blob/main/bug-triage/introduction.md) for more information!
```
(edited)

Reviewer 4: Looks great!

Reviewer: I'm not sure about the "speedy" change...

Reviewer 4: It's the adverb of "speedy", though I missed it should be "seamlessly and speedily", but could be correct to use the adjectives as well if the meaning is that the chunks are speedy and seamless rather than the work

I.e. if it's *making* that is modified, it should be "seamlessly and speedily", and if it's *chunks* it should be "seamless and speedy"

Reviewer 2: "seamless and speedy" are both adjectives, which means they would refer to "_fixes_", which doesn't make much sense.
By converting them to adverbs, they would refer to the act of "_making chunks of fixes_". @AThousandShips only suggested "speedily", but she most likely intended "seamlessly and speedily"

Reviewer: Thanks, yeah it makes sense with seamless being changed too :)

The abstract method PR [GH-106409](https://github.com/godotengine/godot/pull/106409) comes courtesy of [Danil Alexeev](https://github.com/dalexeev), and acts as a direct continuation to the abstract classes introduced in our [previous blog post](/article/dev-snapshot-godot-4-5-dev-5/#gdscript-abstract-classes). By prepending the `abstract` keyword to a function, it will be marked for explicit override by child classes.

```gdscript
abstract class Item:
	abstract func get_name() -> String

	func use() -> void:
		print("Character used %s." % get_name())

class HealingPotion extends Item:
	func get_name() -> String:
		return "Healing Potion"

func _ready() -> void:
	var potion := HealingPotion.new()
	potion.use() # Prints `Character used Healing Potion.`
```

> Danil isn't done yet though, as the variadic PR [GH-82808](https://github.com/godotengine/godot/pull/82808) comes from him as well! In programming languages, variadic arguments allow functions to accept a flexible number of input parameters. This allows turning the final argument of a function into an array that's called as if it were a seqeuence.

Reviewer: ```suggestion
Danil isn't done yet though, as the variadic PR [GH-82808](https://github.com/godotengine/godot/pull/82808) comes from him as well! In programming languages, variadic arguments allow functions to accept a flexible number of input parameters. This allows turning the final argument of a function into an array that is called as if it were a sequence.
```

---

*The cover illustration is from* [**Rift Riff**](https://store.steampowered.com/app/2800900/Rift_Riff/?curator_clanid=41324400), *a tower defense game whose layered combat isn't compromised by its variety of forgiving mechanics! You can buy the game [on Steam](https://store.steampowered.com/app/2800900/Rift_Riff/?curator_clanid=41324400), and follow the developer on [Bluesky](https://bsky.app/profile/adriaan.games).*

## Highlights

- [New in Beta 1!](#new-in-beta-1)
- [Breaking changes](#breaking-changes)
- [Core](#core)
- [Editor](#editor)
- [GDScript](#gdscript)
- [GUI](#gui)
- [Internationalization](#internationalization)
- [Navigation](#navigation)
- [Physics](#physics)
- [Platforms](#platforms)
- [Rendering and shaders](#rendering-and-shaders)

### New in Beta 1!
Much like our [4.4 beta 1 blog post](/article/dev-snapshot-godot-4-4-beta-1/#new-in-beta-1), we have many new features to showcase that managed to squeeze in before the feature freeze! Similarly, most of these are somewhat experimental compared to the more traditional highlights later on in the article, so please let us know if any issues are encountered.

#### Stencil support for spatial materials

> With a PR in the making for nearly 2 years [GH-80710](https://github.com/godotengine/godot/pull/80710), [apples](https://github.com/apples) delivers the long-awaited stencil support to Godot! This new shader functionality is supported on all of our rendering backends, and will allow our users to perform entirely new techniques with the power of **depth**.

Reviewer: ```suggestion
With a PR in the making for nearly 2 years [GH-80710](https://github.com/godotengine/godot/pull/80710), [apples](https://github.com/apples) has brought a long-awaited stencil support to Godot! This new shader functionality is supported on all of our rendering backends, and will allow our users to perform entirely new techniques with the power of **depth**.
```
Mainly to remove the awkward use of present tense.

```gdscript
func f(a: int, b: int = 0, ...args: Array):
	prints(a, b, args)

func _ready() -> void:
	f(1)
	f(1, 2)
	f(1, 2, 3)
	f(1, 2, 3, 4)
	f(1, 2, 3, 4, 5)

# Output:
# 1 0 []
# 1 2 []
# 1 2 [3]
# 1 2 [3, 4]
# 1 2 [3, 4, 5]
```

<img src="/storage/blog/dev-snapshot-godot-4-5-beta-1/variadic-documentation.webp" alt="Variadic documentation"/>

### Breaking changes
We try to minimize breaking changes, but sometimes they are necessary in order to fix high priority issues. Where we do break compatibility, we do our best to make sure that the changes are minimal and require few changes in user projects.

You can find a list of such issues by filtering the merged PRs in the 4.5 milestone with the [`breaks compat` label](https://github.com/godotengine/godot/issues?q=milestone%3A4.5%20is%3Amerged%20label%3A%22breaks%20compat%22). Here are some which are worth being aware of:

- Tilemap physics are now handled in chunks. Only affects `get_coords_for_body_rid`, as now a single body can cover multiple cells. ([GH-102662](https://github.com/godotengine/godot/pull/102662))

> - Internal nodes no longer duplicated. Only affects users deliberately utilizing internal nodes, which shouldn't happen in the first place but hey. ([GH-89442](https://github.com/godotengine/godot/pull/89442))

Reviewer: ```suggestion
- Internal nodes are no longer duplicated. Only affects users deliberately utilizing internal nodes, which shouldn't happen in the first place but hey. ([GH-89442](https://github.com/godotengine/godot/pull/89442))
```

Reviewer 2: It was decided to awkwardly expose internal nodes back in 4.0 (as an `add_child` parameter), so "_which shouldn't happen in the first place_" is a bit misleading.

###### I say "awkwardly" out of personal bias against exposing heavily technical things for pretty much no reason other than making everything look more complicated.

Reviewer 4: Them being duplicated was an oversight (with various hacky half-solutions to fix it), but the wording here seems to imply users aren't supposed to use internal nodes which is wrong

In a similar vein, the ability to manipulate a group of controls simultaneously has become much easier thanks to [Delsin-Yu](https://github.com/Delsin-Yu)'s PR with a focus on recursive control across child controls ([GH-97495](https://github.com/godotengine/godot/pull/97495)). This implements new properties for recursively disabling `Focus Mode` and `Mouse Filter`, meaning that the ability to select and interact with child controls becomes far more intuitive, and allows for explicit overrides if desired.

<video autoplay loop muted playsinline>
  <source src="/storage/blog/godot-4-5-beta-1/recursive-focus.webm?1" type="video/webm">
</video>

> There have also been a few improvements to `SVGTexture` in 4.5, such as the [dedicated importer](https://github.com/godotengine/godot/pull/105655). However, as it pertains to GUI, the biggest improvement would be [bruvzg](https://github.com/bruvzg)'s implementation of auto-scalable SVGs ([GH-105375](https://github.com/godotengine/godot/pull/105375)). This will allow users to explicitly oversample fonts and raster graphics, resulting a much crisper image overall.

Reviewer 5: `SVGTexture` is itself a new class in 4.5, implemented by the second linked PR.

In a similar vein, the ability to manipulate a group of controls simultaneously has become much easier thanks to [Delsin-Yu](https://github.com/Delsin-Yu)'s PR with a focus on recursive control across child controls ([GH-97495](https://github.com/godotengine/godot/pull/97495)). This implements new properties for recursively disabling `Focus Mode` and `Mouse Filter`, meaning that the ability to select and interact with child controls becomes far more intuitive, and allows for explicit overrides if desired.

<video autoplay loop muted playsinline>
  <source src="/storage/blog/godot-4-5-beta-1/recursive-focus.webm?1" type="video/webm">
</video>

> There's also the newly added `SVGTexture`, implemented by our very own [bruvzg](https://github.com/bruvzg) in [GH-105375](https://github.com/godotengine/godot/pull/105375), allowing for rastorization of SVG files directly. However, when initially showcased, this process was fairly long-winded, as the main uses were internal. This has since been remedied with a dedicated importer in [GH-105655](https://github.com/godotengine/godot/pull/105655), making rasterization much more accessible to all users.

Reviewer: ```suggestion
There's also the newly added `SVGTexture`, implemented by our very own [bruvzg](https://github.com/bruvzg) in [GH-105375](https://github.com/godotengine/godot/pull/105375), allowing for rasterization of SVG files directly. However, when initially showcased, this process was fairly long-winded, as the main uses were internal. This has since been remedied with a dedicated importer in [GH-105655](https://github.com/godotengine/godot/pull/105655), making rasterization much more accessible to all users.
```

<video autoplay loop muted playsinline>
  <source src="/storage/blog/dev-snapshot-godot-4-5-dev-4/3d-fti-scenetree.webm?1" type="video/webm">
</video>

And more:
-  Chunk tilemap physics [GH-102662](https://github.com/godotengine/godot/pull/102662)

### Platforms
#### Android
The Android editor experience has become significantly improved thanks to the implementation of `TouchActionsPanel`. While this is technically something any touch device can take advantage of, [Anish Mishra](https://github.com/syntaxerror247)'s PR [GH-100339](https://github.com/godotengine/godot/pull/100339) was explicitly created with the Android editor in mind. `TouchActionsPanel` comes equipped with common action buttons (save, undo, redo, etc.), which simulate actions like `ui_undo` and `ui_redo` via pre-existing shortcuts. This was implemented early on in the 4.5 development cycle, giving our maintainers a stable and versatile foundation for future touch-based actions.

<video autoplay loop muted playsinline>

>   <source src="/storage/blog/godot-4-5-beta-1/touch-actions-panel?1" type="video/webm">

Reviewer 6: There have been a few follow-up PRs (https://github.com/godotengine/godot/pull/105015, https://github.com/godotengine/godot/pull/105140 and https://github.com/godotengine/godot/pull/105518) that improved the panel a lot, so this video is pretty outdated now. It’d be better to replace it with a more recent one.

Reviewer 6: https://github.com/user-attachments/assets/d7faec8d-9a45-4de6-bb15-cf0c763f3206

- Implement motion vectors in mobile renderer ([GH-100283](https://github.com/godotengine/godot/pull/100283))
- Fix LightmapGI shadow leaks ([GH-107254](https://github.com/godotengine/godot/pull/107254))

### XR
- Fragment density map support ([GH-99551](https://github.com/godotengine/godot/pull/99551))

## Changelog

**119 contributors** submitted **372 fixes** for this release. See our [**interactive changelog**](https://godotengine.github.io/godot-interactive-changelog/#4.5-beta1) for the complete list of changes since the previous 4.5-dev5 snapshot. You can also review [all changes included in 4.5](https://godotengine.github.io/godot-interactive-changelog/#4.5) compared to the previous 4.4 feature release.

This release is built from commit [`46c495ca2`](https://github.com/godotengine/godot/commit/46c495ca21f40f57a7fb9c7cde6143738f1652d4).

## Downloads

{% include articles/download_card.html version="4.5" release="beta1" article=page %}

**Standard build** includes support for GDScript and GDExtension.

**.NET build** (marked as `mono`) includes support for C#, as well as GDScript and GDExtension.

{% include articles/prerelease_notice.html %}

> ## Known issues

Reviewer 6: Lets also add https://github.com/godotengine/godot/issues/107535 to known issues.

Reviewer 7: I think https://github.com/godotengine/godot/issues/107438 also might be worth a mention here as it results in constant errors in editor under Wayland and is a regression from 4.5-dev5.

### Core
Changes to the core of the engine require significantly more scrutiny than other parts of the engine; this comes down to how critical and foundational virtually every single piece of code proves to be. This makes it all the more impressive that [bruvzg](https://github.com/bruvzg) managed to implement [AccessKit](https://github.com/AccessKit/accesskit) support in such an integral manner ([GH-76829](https://github.com/godotengine/godot/pull/76829)). Additionally, special thanks to [Lukas Tenbrink](https://github.com/Ivorforce), a new addition to the core team who has been [contributing](https://github.com/godotengine/godot/pull/103708) [nonstop](https://github.com/godotengine/godot/pull/106996) [improvements](https://github.com/godotengine/godot/pull/104381) to ensure optimal performance for developers and maintainers alike.

And more:
- Overhaul resource duplication. ([GH-75950](https://github.com/godotengine/godot/pull/75950))

> ### Documentation

Reviewer 2: Towards the end of this section, if I may suggest, bringing up the [Godot-docs](https://github.com/godotengine/godot-docs) repository (maybe class reference too in general?) and generally pointing out the work from individual people that goes behind ensuring that the documentation is up-to-date with the new features and additions.

- 2D & 3D Navigation region and link updates are now asynchronous. (2D: [GH-107381](https://github.com/godotengine/godot/pull/107381), 3D: [GH-106670](https://github.com/godotengine/godot/pull/106670))
- `NavigationServer2D` avoidance callbacks changed from `Vector3` to `Vector2`. ([GH-107256](https://github.com/godotengine/godot/pull/107256))
- `CONTEXT_SLOT_FILESYSTEM_CREATE` will now always use the base directory. No longer recieves a full path when called. ([GH-107085](https://github.com/godotengine/godot/pull/107085))
- Removed the `gradle_build`/`compress_native_libraries` export option. With Android builds now supporting 16kb pages, the native libraries are now required to be uncompressed. ([GH-106359](https://github.com/godotengine/godot/pull/106359))
- "Areas Detect Static Bodies" setting removed from Jolt Physics. ([GH-105746](https://github.com/godotengine/godot/pull/105746))
- `set_scope` removed from `JSONRPC`. Manual method registration is now required via `set_method`. ([GH-104890](https://github.com/godotengine/godot/pull/104890))

### Animation

> This might be a little more technical than usual, but the work [Tokage](https://github.com/TokageItLab) put into implementing `BoneConstraint3D` warrants a highlight all the same ([GH-100984](https://github.com/godotengine/godot/pull/100984)). With this new class, users will be able to specify two bones simultaneously, opening the door for more natural movements and poses.

Reviewer 8: ```suggestion
This might be a little more technical than usual, but the work [Tokage](https://github.com/TokageItLab) put into implementing `BoneConstraint3D` warrants a highlight all the same ([GH-100984](https://github.com/godotengine/godot/pull/100984)). With this new class, users will be able to bind bone to the another bone, opening the door for more natural movements and poses.
```

Considering that there is a clear master-servant relationship, other terms from `simultaneously` may be more appropriate.

Reviewer 4: Should be:
```suggestion
This might be a little more technical than usual, but the work [Tokage](https://github.com/TokageItLab) put into implementing `BoneConstraint3D` warrants a highlight all the same ([GH-100984](https://github.com/godotengine/godot/pull/100984)). With this new class, users will be able to bind a bone to another bone, opening the door for more natural movements and poses.
```

Reviewer 9: Queued locally in my changes.

<video autoplay loop muted playsinline>
  <source src="/storage/blog/godot-4-5-beta-1/touch-actions-panel?1" type="video/webm">
</video>

And more:
- Add CameraFeed support. ([GH-106094](https://github.com/godotengine/godot/pull/106094))
- Bump minimum supported SDK to 24. ([GH-106148](https://github.com/godotengine/godot/pull/106148))
- Add support for 16KB page sizes. ([GH-106358](https://github.com/godotengine/godot/pull/106358))

#### Linux
[Riteo](https://github.com/Riteo) has been on a quest to turn [Wayland](https://wayland.freedesktop.org/) into a true replacement for the X11 protocol. The biggest reason to not switch over, lack of native sub-windows, is no longer a concern thanks to [GH-101774](https://github.com/godotengine/godot/pull/101774). Now users can truly take advantage of multi-window output, regardless of display protocol.

<img src="/storage/blog/dev-snapshot-godot-4-5-dev-2/wayland-sub-window.webp" alt="Multi-Window output on Wayland"/>

#### macOS

> It took a while for embedded window support to come to macOS, as this OS does not allow the kind of window manipulation that Windows and Linux use for game window embedding. Instead, macOS utilizes an inter-process communication approach where the framebuffer is sent from the game process (which performs off-screen rendering) to the editor window, which also handles input events to the game process. Although it's much more complex, [Stuart Carnie](https://github.com/stuartcarnie) used a more robust approach that doesn't rely on any window management hacks ([GH-105884](https://github.com/godotengine/godot/pull/105884)). Not having to worry about edge-cases or low-level shenanigans is appealing in its own right, so this approach may be ported later to Windows/Linux in a future release to make game window embedding much more reliable overall.

Reviewer 4: ```suggestion
It took a while for embedded window support to come to macOS, as this OS does not allow the kind of window manipulation that Windows and Linux use for game window embedding. Instead, macOS utilizes an inter-process communication approach where the framebuffer is sent from the game process (which performs off-screen rendering) to the editor window, which also handles passing input events to the game process. Although it's much more complex, [Stuart Carnie](https://github.com/stuartcarnie) used a more robust approach that doesn't rely on any window management hacks ([GH-105884](https://github.com/godotengine/godot/pull/105884)). Not having to worry about edge-cases or low-level shenanigans is appealing in its own right, so this approach may be ported later to Windows/Linux in a future release to make game window embedding much more reliable overall.
```

#### visionOS
The Apple XR environment, visionOS, comes to Godot Engine in 4.5! We're already committed to making great strides in XR featuresets and support, but this brings us closer to a platform-independent pipeline similar to what we already have with traditional platforms. It might be unconventional to imagine using an engine editor within an XR environment, but it's one which we've supported for Meta Quest and OpenXR; now Apple users can join the fun.

#### Web
The upcoming performance boost from SIMD is so impressive that we've made a [dedicated article](/article/upcoming-serious-web-performance-boost/) to showcase exactly what kind of benefits users can expect. The short version is that [Adam Scott](https://github.com/adamscott) created a PR which changed only a single compiler flag ([GH-106319](https://github.com/godotengine/godot/pull/106319)), but one which caused universal improvements for web applications and editor workflow. Check out our article for the long version.

#### Windows

> Support for Windows 7/8.1 will be dropped starting with 4.5 ([GH-106959](https://github.com/godotengine/godot/pull/106959)). The call to remove Windows 8.1 wasn't a difficult one; it's been EOL for over half a decade, had extended support end over two years ago, and was inherently unpopular to a point that online survey tools for OSes outright omit it. Windows 7 is comparatively more contentious, but builds for it were already in a broken state ever since the introduction of the aforementioned AccessKit; combine that with Windows 7 recently celebrating its 10-year anniversary of being EOL, as well as active coverage of this OS being estimated at one-tenth of a percentage point, and its removal was ultimately cemented. This had the benefit of **significantly** cleaning up the codebase for our Windows-specific files, which often had to rely on dummy includes to account for Windows 7 specifically, and opens the door for more modern APIs to be integrated for our Windows builds.

Reviewer 4: ```suggestion
Support for Windows 7/8.1 will be dropped starting with 4.5 ([GH-106959](https://github.com/godotengine/godot/pull/106959)). The call to remove Windows 8.1 wasn't a difficult one; it's been EOL for over half a decade, had extended support end over two years ago, and was inherently unpopular to the point that online survey tools for OSes outright omit it. Windows 7 is comparatively more contentious, but builds for it were already in a broken state ever since the introduction of the aforementioned AccessKit; combine that with Windows 7 recently celebrating its 10-year anniversary of being EOL, as well as active coverage of this OS being estimated at one-tenth of a percentage point, and its removal was ultimately cemented. This had the benefit of **significantly** cleaning up the codebase for our Windows-specific files, which often had to rely on dummy includes to account for Windows 7 specifically, and opens the door for more modern APIs to be integrated for our Windows builds.
```

The abstract method PR [GH-106409](https://github.com/godotengine/godot/pull/106409) comes courtesy of [Danil Alexeev](https://github.com/dalexeev), and acts as a direct continuation to the abstract classes introduced in our [previous blog post](/article/dev-snapshot-godot-4-5-dev-5/#gdscript-abstract-classes). By prepending the `abstract` keyword to a function, it will be marked for explicit override by child classes.

```gdscript
abstract class Item:
	abstract func get_name() -> String

	func use() -> void:
		print("Character used %s." % get_name())

class HealingPotion extends Item:
	func get_name() -> String:
		return "Healing Potion"

func _ready() -> void:
	var potion := HealingPotion.new()
	potion.use() # Prints `Character used Healing Potion.`
```

> **Note:** The GDScript is planning to change the `abstract` keyword to an `@abstract` annotation during the 4.5 beta phase. So while the syntax is with a keyword in this beta 1 build, the example above will change slightly in a future update.

Reviewer 10: ```suggestion
**Note:** The GDScript team is planning to change the `abstract` keyword to an `@abstract` annotation during the 4.5 beta phase. So while the syntax is with a keyword in this beta 1 build, the example above will change slightly in a future update.
```

This might be a little more technical than usual, but the work [Tokage](https://github.com/TokageItLab) put into implementing `BoneConstraint3D` warrants a highlight all the same ([GH-100984](https://github.com/godotengine/godot/pull/100984)). With this new class, users will be able to bind a bone to another bone, opening the door for more natural movements and poses.

**Twist:**

|                                                              Disabled                                                               |                                                              Enabled                                                              |
| :---------------------------------------------------------------------------------------------------------------------------------: | :-------------------------------------------------------------------------------------------------------------------------------: |
| <img src="/storage/blog/dev-snapshot-godot-4-5-beta-1/constraint-twist-disabled.webp" alt="Constraint twist disabled" width="350"/> | <img src="/storage/blog/dev-snapshot-godot-4-5-beta-1/constraint-twist-enabled.webp" alt="Constraint twist enabled" width="350"/> |

**Bend:**

> <video autoplay loop muted playsinline>

Reviewer: ```suggestion
<video autoplay loop muted playsinline title="An example of bone constraints on a bent knee">
```
And so on.

Additionally, special thanks to [Lukas Tenbrink](https://github.com/Ivorforce), a new addition to the core team who has been [contributing](https://github.com/godotengine/godot/pull/103708) [nonstop](https://github.com/godotengine/godot/pull/106996) [improvements](https://github.com/godotengine/godot/pull/104381) to ensure optimal performance for developers and maintainers alike.

And more:
- Add `Node.get_orphan_node_ids`, edit `Node.print_orphan_nodes`. ([GH-83757](https://github.com/godotengine/godot/pull/83757))
- Don't duplicate internal nodes. ([GH-89442](https://github.com/godotengine/godot/pull/89442))
- Use Grisu2 algorithm in `String::num_scientific` to fix serializing. ([GH-98750](https://github.com/godotengine/godot/pull/98750))
- Add `scene_changed` signal to SceneTree. ([GH-102986](https://github.com/godotengine/godot/pull/102986))
- Complete build profile feature to properly detect options that can be disabled (reducing binary size). ([GH-103719](https://github.com/godotengine/godot/pull/103719))
- Add thread safety to Object signals. ([GH-105453](https://github.com/godotengine/godot/pull/105453))

### Documentation

> It's not often that we have the opportunity to cover documentation changes in these highlights, as those changes are usually fairly low-key. This isn't to say that changes aren't happening; on the contrary, it's one of the single most active areas of our GitHub! That's because it often goes beyond our main repo, with changes needing to be synchronized in [godot-docs](https://github.com/godotengine/godot-docs). Special shoutouts to [Mickeon](https://github.com/Mickeon) for the class reference, and [Matthew](https://github.com/skyace65), [tetrapod](https://github.com/tetrapod00), [A Thousand Ships](https://github.com/AThousandShips), [Max Hilbrunner](https://github.com/mhilbrunner) and [Hugo Locurcio](https://github.com/Calinou) for the online docs, for taking on the lion's share of pull requests *and* reviews.

Reviewer: Should it be "A Thousand Ships" or "AThousandShips"? You used "AThousandShips" earlier.

Reviewer 9: I usually go with what's on the contributor's GH profile as their display name, so in this case "A Thousand Ships" indeed. I fixed the other occurrences.

	prints(a, b, args)

func _ready() -> void:
	f(1)
	f(1, 2)
	f(1, 2, 3)
	f(1, 2, 3, 4)
	f(1, 2, 3, 4, 5)

# Output:
# 1 0 []
# 1 2 []
# 1 2 [3]
# 1 2 [3, 4]
# 1 2 [3, 4, 5]
```

<img src="/storage/blog/dev-snapshot-godot-4-5-beta-1/variadic-documentation.webp" alt="Variadic documentation"/>

GDScript also saw a lot of usability improvements for the script editor, with notably fixes to code completion and highlighting:
- Inline color pickers. ([GH-105724](https://github.com/godotengine/godot/pull/105724))
- Highlight warning lines. ([GH-105724](https://github.com/godotengine/godot/pull/105724))
- Don't add parenthesis when expecting `Callable`. ([GH-96375](https://github.com/godotengine/godot/pull/96375))
- Autocompletion for `@export_tool_button`s. ([GH-105081](https://github.com/godotengine/godot/pull/105081))
- Autocompletion for user methods. ([GH-106198](https://github.com/godotengine/godot/pull/106198))
- Highlighting script members like native ones. ([GH-74393](https://github.com/godotengine/godot/pull/74393))
- Code completion for overridden user-defined methods. ([GH-106198](https://github.com/godotengine/godot/pull/106198))

### GUI

> Here at the Godot Foundation, accessibililty is an absolute top-priority. The road to making an experience available to anyone regardless of their circumstances isn't an easy one, but it's a road that all developers are obligated to take. To that end, our resident tech guru [bruvzg](https://github.com/bruvzg) tackled absolutely Herculean task of integrating [AccessKit](https://github.com/AccessKit/accesskit) to Godot as a whole ([GH-76829](https://github.com/godotengine/godot/pull/76829)). With this in place, screenreader support is now built into the very core of the engine. All our supported desktop platforms offer fully uncompromised support, as the bindings are already in place and well-tested. When other platforms follow suit, we will ensure support to the absolute best of our ability.

Reviewer: ```suggestion
Here at the Godot Foundation, accessibililty is an absolute top-priority. The road to making an experience available to anyone regardless of their circumstances isn't an easy one, but it's a road that all developers are obligated to take. To that end, our resident tech guru [bruvzg](https://github.com/bruvzg) tackled the absolutely Herculean task of integrating [AccessKit](https://github.com/AccessKit/accesskit) to Godot as a whole ([GH-76829](https://github.com/godotengine/godot/pull/76829)). With this in place, screenreader support is now built into the very core of the engine. All our supported desktop platforms offer fully uncompromised support, as the bindings are already in place and well-tested. When other platforms follow suit, we will ensure support to the absolute best of our ability.
```

- Optimize Mobile renderer by using FP16 explicitly. ([GH-107119](https://github.com/godotengine/godot/pull/107119))
- Fix LightmapGI shadow leaks. ([GH-107254](https://github.com/godotengine/godot/pull/107254))

### XR

The XR team has been doing a [lot of groundwork](https://github.com/godotengine/godot/pulls?q=is%3Amerged+is%3Apr+label%3Atopic%3Axr+milestone%3A4.5+label%3Aenhancement) to support more OpenXR extensions. While it's separate from the main Godot release cycle, the Godot OpenXR Vendors had a [4.0.0 release](https://github.com/GodotVR/godot_openxr_vendors/releases/tag/4.0.0-stable) in April which is worth checking out! [Matthieu Bucchianeri](https://github.com/mbucchia) had a massive first contribution merged to implement support for the Direct3D 12 OpenXR backend ([GH-104207](https://github.com/godotengine/godot/pull/104207)).

[Darío](https://github.com/DarioSamo) and [Logan Lang](https://github.com/devloglogan)'s implementations of respectively fragment density maps and motion vectors for the Mobile renderer also pave the way for better rendering and supporting more extensions for XR.

> While those are not merged yet, you can keep an eye on [Bastiaan Olij](https://github.com/BastiaanOlij)'s cutting edge implementations for the OpenXR Spatial Entities ([GH-107391](https://github.com/godotengine/godot/pull/107391)) and Render Model ([GH-107388](https://github.com/godotengine/godot/pull/107388)) extensions, building upon the freshly released OpenXR SDK 1.1.49. [Fredia Huya-Kouadio](https://github.com/m4gr3d) is also putting finishing touches on the support for running hybrid apps from the Godot XR editor ([GH-103972](https://github.com/godotengine/godot/pull/103972)).

Reviewer 11: "Those" makes it sound like it's talking about the stuff from the previous paragraph, rather than the current paragraph:

```suggestion
While these are not merged yet, you can keep an eye on [Bastiaan Olij](https://github.com/BastiaanOlij)'s cutting edge implementations for the OpenXR Spatial Entities ([GH-107391](https://github.com/godotengine/godot/pull/107391)) and Render Model ([GH-107388](https://github.com/godotengine/godot/pull/107388)) extensions, building upon the freshly released OpenXR SDK 1.1.49. [Fredia Huya-Kouadio](https://github.com/m4gr3d) is also putting finishing touches on the support for running hybrid apps from the Godot XR editor ([GH-103972](https://github.com/godotengine/godot/pull/103972)).
```

Reviewer 9: Thanks, good catch!

### C\#

> First-time contributor [Justin Sasso](https://github.com/atlasapplications) kicks things off with `linux-bionic` RID export support ([GH-97908](https://github.com/godotengine/godot/pull/97908)). For those that don't speak buildsystem, this enables NativeAOT on Android! For those that don't speak nerd, "NativeAOT" refers to the ability for .NET applications to compile directly to a device's native code, bypassing the need for the .NET runtime entirely. NativeAOT apps have the benefit of significantly faster startup and smaller memory footprints, which are both very welcome additions for mobile devices.

Reviewer: Best not say "nerd"!

- Code completion for overridden user-defined methods. ([GH-106198](https://github.com/godotengine/godot/pull/106198))

### GUI

> Here at the Godot Foundation, accessibililty is an absolute top-priority. The road to making an experience available to anyone regardless of their circumstances isn't an easy one, but it's a road that all developers are obligated to take. To that end, our resident tech guru [bruvzg](https://github.com/bruvzg) tackled absolutely Herculean task of integrating [AccessKit](https://github.com/AccessKit/accesskit) to Godot as a whole ([GH-76829](https://github.com/godotengine/godot/pull/76829)). With this in place, screenreader support is now built into the very core of the engine. All our supported desktop platforms offer fully uncompromised support, as the bindings are already in place and well-tested; when other platforms follow suit, we will ensure support to the absolute best of our ability.

Reviewer: ```suggestion
Here at the Godot Foundation, accessibililty is an absolute top-priority. The road to making an experience available to anyone regardless of their circumstances isn't an easy one, but it's a road that all developers are obligated to take. To that end, our resident tech guru [bruvzg](https://github.com/bruvzg) tackled the absolutely Herculean task of integrating [AccessKit](https://github.com/AccessKit/accesskit) to Godot as a whole ([GH-76829](https://github.com/godotengine/godot/pull/76829)). With this in place, screenreader support is now built into the very core of the engine. All our supported desktop platforms offer fully uncompromised support, as the bindings are already in place and well-tested. When other platforms follow suit, we will ensure support to the absolute best of our ability.
```
